server/src/main/java/org/elasticsearch/index/codec/tsdb/es819/ES819TSDBDocValuesConsumer.java

```diff
@@ -373,75 +389,239 @@ public void addBinaryField(FieldInfo field, DocValuesProducer valuesProducer) th
+        int maxNumDocsInAnyBlock = 0;
+
+        final BlockMetadataAccumulator blockMetaAcc;
+
+        CompressedBinaryBlockWriter(BinaryDVCompressionMode compressionMode) throws IOException {
+            this.compressor = compressionMode.compressionMode().newCompressor();
+            long blockAddressesStart = data.getFilePointer();
+            this.blockMetaAcc = new BlockMetadataAccumulator(state.directory, state.context, data, blockAddressesStart);
+        }
+
+        @Override
+        public void addDoc(BytesRef v) throws IOException {
+            block = ArrayUtil.grow(block, uncompressedBlockLength + v.length);
+            System.arraycopy(v.bytes, v.offset, block, uncompressedBlockLength, v.length);
+            uncompressedBlockLength += v.length;
+
+            numDocsInCurrentBlock++;
+            docRanges = ArrayUtil.grow(docRanges, numDocsInCurrentBlock + 1); // need one extra since writing start for next block
+            docRanges[numDocsInCurrentBlock] = uncompressedBlockLength;
+
+            if (uncompressedBlockLength >= BLOCK_BYTES_THRESHOLD || numDocsInCurrentBlock >= BLOCK_COUNT_THRESHOLD) {
+                flushData();
+            }
+        }
+
+        @Override
+        public void flushData() throws IOException {
+            if (numDocsInCurrentBlock == 0) {
+                return;
+            }
+
+            totalChunks++;
+            long thisBlockStartPointer = data.getFilePointer();
+
+            // write length of string data
+            data.writeInt(uncompressedBlockLength);
+
+            maxUncompressedBlockLength = Math.max(maxUncompressedBlockLength, uncompressedBlockLength);
+            maxNumDocsInAnyBlock = Math.max(maxNumDocsInAnyBlock, numDocsInCurrentBlock);
+
+            compressOffsets(data, numDocsInCurrentBlock);
+            compress(block, uncompressedBlockLength, data);
+
+            long blockLenBytes = data.getFilePointer() - thisBlockStartPointer;
+            blockMetaAcc.addDoc(numDocsInCurrentBlock, blockLenBytes);
+            numDocsInCurrentBlock = uncompressedBlockLength = 0;
+        }
+
+        void compressOffsets(DataOutput output, int numDocsInCurrentBlock) throws IOException {
+            int batchStart = 0;
+            int numOffsets = numDocsInCurrentBlock + 1;
+            while (batchStart < numOffsets) {
+                int batchLength = Math.min(numOffsets - batchStart, NUMERIC_BLOCK_SIZE);
+                for (int i = 0; i < batchLength; i++) {
+                    docRangesBuffer[i] = docRanges[batchStart + i];
+                }
+                if (batchLength < docRangesBuffer.length) {
+                    Arrays.fill(docRangesBuffer, batchLength, docRangesBuffer.length, 0);
+                }
+                encoder.encode(docRangesBuffer, output);
+                batchStart += batchLength;
+            }
+        }
+
+        void compress(byte[] data, int uncompressedLength, DataOutput output) throws IOException {
```

Comment:
If we use Zstd directly, should we also handle cases where compression does not reduce storage and store the raw bytes instead?